extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/base/ModeledDirectoryObjectService.java

```diff
@@ -427,16 +427,22 @@ public InternalType retrieveObject(ModeledAuthenticatedUser user,
     protected Collection<ObjectPermissionModel> getImplicitPermissions(ModeledAuthenticatedUser user,
             ModelType model) {
         
+        // Get the user model and check for an entity ID.
+        UserModel userModel = user.getUser().getModel();
+        Integer entityId = userModel.getEntityID();
+        if (entityId == null)
+            return Collections.emptyList();
```

Comment:
I'm still unclear on the circumstances that would cause `entityId` to be `null`. The entity serves as the permission-receiving base of either a user or a group, and referential integrity constraints guarantee that neither a user nor a group can be inserted without a corresponding entity.

For users, the entity is currently automatically inserted during creation of the user within `beforeCreate()` (invoked by `createObject()`):

https://github.com/apache/guacamole-client/blob/e1f95b87636705931fff50fd37598bb63a26ae48/extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/UserService.java#L253-L254

Is there something going on which leads to this check being required? If so, and `entityId` can indeed be `null` here, how are things working without an entity?

Reply:
Okay, I think I see what you're getting at.  Basically...
- `createObject()` requires two parameters - the user creating the object, and the object being created.
- In `JDBCAuthenticationService` I make the call to `userService.createObject()` to create the user.  The second argument for that call is obviously the user record that is being automatically created, here.  The first argument (as currently implemented) is a new `PrivilegedModeledAuthenticatedUser` based on the user that was authenticated and is currently being added.  So, essentially, we're giving the user that is currently being added the privilege to create their own account in the database.
- However, because that user doesn't already exist (presumably _at the time of_ the `new PrivilegedModeledAuthenticatedUser() call), there is no entity, so the `entityId` field is null, resulting in a constraint violation when trying to insert those implicit privileges.
- Thus, the only implicit privilege that actually ever gets created with these checks is the one for the user to `READ` their own account.  Which I think is actually okay and desirable - otherwise, the user will end up with full privileges on their own account, which might be okay but also might be more than what the sysadmin desires, and I'm all about least privilege.

Let me know if there's something different I should do, here:
- Move the `new PrivilegedModeledAuthenticatedUser()` to a different place.
- Use the `userContext.getPrivileged()` instead of manually creating that user.
- Use a different based account for the `new PrivilegedModeledAuthenticatedUser()` call such that an entity exists.
- Something else?

Reply:
Ah, OK ... I think I see what you're saying.

For users, `getImplicitPermissions()` contains essentially two distinct sets of permissions:

1. Those from `IMPLICIT_USER_PERMISSIONS`, which are the permissions that a newly-created user should have _for themselves_. This consists only of `READ`, and should successfully be granted to an automatically-created user.
2. Those from `super.getImplicitPermissions()`, which are the permissions granted to the user creating the object. This consists of several admin-level permissions, and in this case both can't and shouldn't be granted:
   * "Can't" because the user creating the object does not exist.
   * "Shouldn't" because automatically-created users shouldn't be magically granted permission to administer themselves.

Is that correct?

Reply:
Yep, exactly.  So, I think the way it's working right now is desirable, as the `IMPLICIT_USER_PERMISSIONS` are granted (user can `READ` self), and the administrative permissions granted by `super.getImplicitPermissions()` are not granted.  I just don't know if everything is set up in the best possible way or if there's a different/better way to accomplish that.

Reply:
I think what you have makes sense, then, but that check on whether `entityId` is `null` should be documented to clarify this need.

I suggest adding something like `isSkeleton()` to `ModeledUser` to abstract away this check. Then the code itself would embody the idea behind what you're checking, and the check within the superclass version of `getImplicitPermissions()` would be that much clearer.

Reply:
Okay, stab has been taken at your suggestion.